File: 2022/day01.pl
use v5.36;

local $/ = "\n\n";
open my $fh, '<', 'input/day01.txt' or die "bad open: $!";

my @cals;

while (my $hunk = <$fh>) {
  my $cur = 0;
  $cur += $_ for split /\n/, $hunk;
  push @cals, $cur;
}

my ($one, $two, $three) = sort { $b <=> $a } @cals;

say 'part 1: ' . $one;
say 'part 2: ' . ($one + $two + $three);

File: 2022/src/bin/day01.rs
use std::collections::BTreeSet;
use std::fs::File;
use std::io::{prelude::*, BufReader};

fn main() -> Result<(), std::io::Error> {
  let reader = BufReader::new(File::open("input/day01.txt")?);

  let mut vals: BTreeSet<u32> = BTreeSet::new();
  let mut cur: u32 = 0;

  for line in reader.lines().map(|l| l.unwrap().trim().to_string()) {
    if line.is_empty() {
      vals.insert(cur);
      cur = 0;
      continue;
    }

    let val: u32 = line.parse().expect("bad int");
    cur += val;
  }

  let top = vals.iter().rev().take(3).collect::<Vec<_>>();

  println!("part 1: {}", top[0]);
  println!("part 2: {}", top.iter().copied().sum::<u32>());

  Ok(())
}
